lxd/instance/drivers/driver_lxc.go

```diff
@@ -1301,6 +1301,48 @@ func (d *lxc) initLXC(config bool) error {
 	return nil
 }
 
+var idmappedStorageMap map[unix.Fsid]idmap.IdmapStorageType = map[unix.Fsid]idmap.IdmapStorageType{}
+var idmappedStorageMapLock sync.Mutex
+
+// IdmappedStorage determines if the container can use idmapped mounts or shiftfs
+func (d *lxc) IdmappedStorage(path string) idmap.IdmapStorageType {
+	var mode idmap.IdmapStorageType = idmap.IdmapStorageNone
+
+	if d.state.OS.Shiftfs {
+		// Fallback to shiftfs.
+		mode = idmap.IdmapStorageShiftfs
+	}
+
+	if !d.state.OS.LXCFeatures["idmapped_mounts"] {
+		return mode
+	}
+
+	buf := &unix.Statfs_t{}
+	err := unix.Statfs(path, buf)
+	if err != nil {
+		// Log error but fallback to shiftfs
+		d.logger.Error("%s - Failed to statfs", err)
```

Comment:
@brauner I'm not sure this is valid, and I think it will result in a runtime panic if it ever gets here. The instance level logger (d.logger) doesn't provide the `Errorf()` function which is what it looks like you were expecting to use here. Instead you should use the contextual logger approach, i.e `d.logger.Error("Failed to statfs", log.Ctx{"err":err})`

Reply:
Ah, good catch.

I'm trying to find out why everything succeeds locally with the same kernel and LXC version but fails on jenkins across the board so I'm doing some tests pushing up to different commits.